From 8c07ecfe89b7399c7db79c9627b5a1b396cb61b0 Mon Sep 17 00:00:00 2001
From: Nick Dimiduk <ndimiduk@apache.org>
Date: Thu, 13 Aug 2015 16:19:26 -0700
Subject: [PATCH 258/295] HBASE-10844 Coprocessor failure during batchmutation
 leaves the memstore datastructs in an inconsistent
 state

Reason: Supportability
Author: Nick Dimiduk
Ref: CDH-31924
---
 .../apache/hadoop/hbase/regionserver/HRegion.java  |    9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java
index 8ee2305..8c42177 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java
@@ -1375,7 +1375,14 @@ public class HRegion implements HeapSize, PropagatingConfigurationObserver { //
 
         // close each store in parallel
         for (final Store store : stores.values()) {
-          assert abort || store.getFlushableSize() == 0 || writestate.readOnly;
+          long flushableSize = store.getFlushableSize();
+          if (!(abort || flushableSize == 0 || writestate.readOnly)) {
+            getRegionServerServices().abort("Assertion failed while closing store "
+                + getRegionInfo().getRegionNameAsString() + " " + store
+                + ". flushableSize expected=0, actual= " + flushableSize
+                + ". Current memstoreSize=" + getMemstoreSize() + ". Maybe a coprocessor "
+                + "operation failed and left the memstore in a partially updated state.", null);
+          }
           completionService
               .submit(new Callable<Pair<byte[], Collection<StoreFile>>>() {
                 @Override
-- 
1.7.9.5

